cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1243,6 +1243,48 @@ class ObjectOutputStream final : public io::OutputStream {
     return Status::OK();
   }
 
+  Future<> CloseAsync() override {
```

Comment:
There's a lot of overlap between this method and `Close`.  Can we create some helper methods so we have something like (pseudo-code)...

```
Future<> CloseAsync() override {
  PrepareClose();
  FlushAsync().Then(DoClose);
}

Status Close() override {
  PrepareClose();
  Flush();
  DoClose();
}
```

Or you could just change `Close` to be `return CloseAsync().status()`

Reply:
Ah whoops, I had meant to make Close() just call CloseAsync(), thanks for catching this.